garage/tf/models/gaussian_mlp_model.py

```diff
@@ -0,0 +1,220 @@
+"""Gaussian MLP Model."""
+import numpy as np
+import tensorflow as tf
+
+from garage.core import Serializable
+from garage.misc import ext
+from garage.misc.overrides import overrides
+from garage.tf.core.networks import mlp, parameter
+from garage.tf.models import Model
+
+
+class GaussianMLPModel(Model, Serializable):
+    """Gaussian MLP Model."""
+
+    def __init__(
+            self,
+            input_dim,
+            output_dim,
+            name="GaussianMLPModel",
+            hidden_sizes=(32, 32),
+            learn_std=True,
+            init_std=1.0,
+            adaptive_std=False,
+            std_share_network=False,
+            std_hidden_sizes=(32, 32),
+            min_std=1e-6,
+            max_std=None,
+            std_hidden_nonlinearity=tf.nn.tanh,
+            hidden_nonlinearity=tf.nn.tanh,
+            output_nonlinearity=None,
+            std_parameterization='exp',
+    ):
+        """
+        Initialize a gaussian mlp model and build the graph.
+
+        :param input_dim: input dimension
+        :param output_dim: output dimension
+        :param name: name of the model
+        :param hidden_sizes: list of sizes for the fully-connected hidden
+          layers
+        :param learn_std: Is std trainable
+        :param init_std: Initial std
+        :param adaptive_std:
+        :param std_share_network:
+        :param std_hidden_sizes: list of sizes for the fully-connected layers
+          for std
+        :param min_std: whether to make sure that the std is at least some
+          threshold value, to avoid numerical issues
+        :param std_hidden_nonlinearity:
+        :param hidden_nonlinearity: nonlinearity used for each hidden layer
+        :param output_nonlinearity: nonlinearity for the output layer
+        :param mean_network: custom network for the output mean
+        :param std_network: custom network for the output log std
+        :param std_parametrization: how the std should be parametrized. There
+          are a few options:
+            - exp: the logarithm of the std will be stored, and applied a
+                exponential transformation
+            - softplus: the std will be computed as log(1+exp(x))
+        :return:
+        """
+        Serializable.quick_init(self, locals())
+        super(GaussianMLPModel, self).__init__()
+
+        self.name = name
+        self._variable_scope = tf.variable_scope(
+            self.name, reuse=tf.AUTO_REUSE)
```

Comment:
Why use tf.AUTO_REUSE here?

Reply:
I assume parameters have unique fullnames (I.e. “variable_scope/name”)

Reply:
I have the same thought for MLP too (each different mlp should have unique fullname)

Reply:
You should not set the reuse scope. Force the outer scope (e.g. the author) to decide which scope behavior to use.

Reply:
To add detail: all parameters internal to this scope still need unique fullnames, but if there is a name collision, tf.AUTO_REUSE can cause TF to silently reuse parameters in a way that the author doesn't expect and can be difficult to debug.

That's why it's not the default in TF and I don't think it should be default here.

If an author wants tf.AUTO_REUSE they can declare that in the enclosing scope.